File: frontend/src/crop.test.ts
import { describe, it, expect } from "vitest";
import {
  cropLayout,
  cropPixels,
  cropToAspect,
  fillsFrame,
  fitMode,
  isEmptyCrop,
  sourceSize,
} from "./crop";
import type { Asset, Clip } from "./types";

const asset = (w: number, h: number) => ({ width: w, height: h }) as Asset;
const clip = (over: Partial<Clip> = {}) => over as Clip;

describe("cropPixels", () => {
  // Mirrors schema.Crop.Pixels — the two are the same arithmetic in two
  // languages, and a divergence shows up as a preview and an export a pixel
  // apart with nothing to explain why.
  it("matches the Go side on the headline case", () => {
    expect(cropPixels({ top: 0.25 }, 1920, 1080)).toEqual({ x: 0, y: 270, w: 1920, h: 810 });
  });

  it("keeps every value even", () => {
    const r = cropPixels({ top: 0.137, left: 0.211, right: 0.073, bottom: 0.019 }, 1913, 1077);
    for (const [k, v] of Object.entries(r)) expect(v % 2, k).toBe(0);
    expect(r.x + r.w).toBeLessThanOrEqual(1913);
    expect(r.y + r.h).toBeLessThanOrEqual(1077);
  });

  // An untrimmed edge starts at zero. Rounding it up to two — which the size
  // rule does, and must — would shift every uncropped axis by a pixel.
  it("leaves an untrimmed origin at zero", () => {
    expect(cropPixels({ bottom: 0.1 }, 1920, 1080).x).toBe(0);
    expect(cropPixels({ right: 0.1 }, 1920, 1080).y).toBe(0);
  });

  it("cannot be dragged down to nothing", () => {
    const r = cropPixels({ top: 0.8, bottom: 0.8, left: 0.95, right: 0.95 }, 1920, 1080);
    expect(r.w).toBeGreaterThanOrEqual(2);
    expect(r.h).toBeGreaterThanOrEqual(2);
    expect(r.w).toBeLessThanOrEqual(1920);
  });
});

describe("sourceSize", () => {
  it("is the picture's size, not the file's", () => {
    expect(sourceSize(asset(1920, 1080), clip({ crop: { top: 0.25 } }))).toEqual({ width: 1920, height: 810 });
    expect(sourceSize(asset(1920, 1080), clip())).toEqual({ width: 1920, height: 1080 });
  });

  // An unprobed asset stays unknown. Turning 0x0 into a confident 2x2 would
  // letterbox such a clip into a speck rather than leaving it alone.
  it("stays undefined for an asset of unknown size", () => {
    expect(sourceSize(asset(0, 0), clip({ crop: { top: 0.2 } }))).toBeUndefined();
    expect(sourceSize(undefined, clip())).toBeUndefined();
  });
});

describe("fitMode", () => {
  // Mirrors prefitFilter and schema.FitCovers: an unset fit letterboxes, full
  // stop. It used to fill wherever the camera was working the clip, and since
  // every screen recording carries cursor effects that cropped a quarter off any
  // recording whose shape did not match the canvas, silently.
  it("letterboxes unless told otherwise", () => {
    expect(fitMode("")).toBe("fit");
    expect(fitMode(undefined)).toBe("fit");
    expect(fitMode("fit")).toBe("fit");
    expect(fitMode("fill")).toBe("fill");
    expect(fitMode("stretch")).toBe("stretch");
  });

  it("reports which modes leave no bars to protect a pan from", () => {
    expect(fillsFrame("")).toBe(false);
    expect(fillsFrame(undefined)).toBe(false);
    expect(fillsFrame("fit")).toBe(false);
    expect(fillsFrame("fill")).toBe(true);
    expect(fillsFrame("stretch")).toBe(true);
  });
});

describe("cropLayout", () => {
  /*
   * The element is sized to the WHOLE source and positioned so the surviving
   * rectangle lands where it belongs, with the parent clipping the rest. One
   * rule for all three fits; object-fit could express two of them and neither
   * with a crop applied.
   */
  it("places an uncropped source exactly as contain would", () => {
    // 16:9 source into a 16:9 box: fills it, no offset.
    const l = cropLayout(undefined, { width: 1920, height: 1080 }, 640, 360, "fit");
    expect(l.window).toEqual({ left: 0, top: 0, width: 640, height: 360 });
    expect(l.media).toEqual({ left: 0, top: 0, width: 640, height: 360 });
  });

  it("letterboxes a source that is the wrong shape", () => {
    // 1:1 source into a 2:1 box: a 180-wide picture centred in 360.
    const l = cropLayout(undefined, { width: 1000, height: 1000 }, 360, 180, "fit");
    expect(l.window.width).toBeCloseTo(180);
    expect(l.window.height).toBeCloseTo(180);
    expect(l.window.left).toBeCloseTo(90);
    expect(l.window.top).toBeCloseTo(0);
  });

  /*
   * The bug this shape exists to prevent, caught by running the real thing.
   *
   * A fitted crop's window is SMALLER than the clip's box. Clipping to the box
   * — which is what a single overflow:hidden parent does — leaves the trimmed
   * edges on screen, filling what should be letterbox bars with exactly the
   * pixels the crop was asked to remove. The crop then appears to do nothing,
   * while the export (which really does drop them) disagrees.
   */
  it("clips to the crop, not to the clip's box, when fitting", () => {
    // A 3456x2234 screen recording with the top quarter cut, into a 16:9 box.
    const l = cropLayout({ top: 0.25 }, { width: 3456, height: 2234 }, 696, 392, "fit");
    // The surviving picture is wider than 16:9, so it letterboxes: the window
    // is shorter than the box and centred in it.
    expect(l.window.height).toBeLessThan(392);
    expect(l.window.top).toBeCloseTo((392 - l.window.height) / 2, 6);
    expect(l.window.width).toBeCloseTo(696, 6);
    // And the removed edge hangs above the window, where it is clipped away.
    expect(l.media.top).toBeLessThan(0);
    expect(l.media.height).toBeGreaterThan(l.window.height);
  });

  /*
   * The other half. Cut the top quarter off a 16:9 recording and fill: the
   * surviving 1920x810 covers a 16:9 box, so it is scaled up and the removed
   * band sits above the window.
   */
  it("covers the box when filling, with the cut edge outside the window", () => {
    const l = cropLayout({ top: 0.25 }, { width: 1920, height: 1080 }, 1920, 1080, "fill");
    const k = 1080 / 810; // scale needed to cover on the short axis
    // The window covers the box and overflows it sideways — which is what fill
    // means, and why the clip's box clips too rather than only the window.
    expect(l.window.height).toBeCloseTo(1080, 6);
    expect(l.window.width).toBeCloseTo(1920 * k, 6);
    expect(l.window.left).toBeCloseTo((1920 - 1920 * k) / 2, 6);
    expect(l.window.top).toBeCloseTo(0, 6);
    // The removed 270px of source, scaled, sits above the window.
    expect(l.media.width / 1920).toBeCloseTo(k, 3);
    expect(l.media.top).toBeCloseTo(-270 * k, 1);
  });

  it("scales the axes apart only when stretching", () => {
    const l = cropLayout({ top: 0.25 }, { width: 1920, height: 1080 }, 1920, 1080, "stretch");
    // The surviving 1920x810 is pulled to the full 1920x1080 box.
    expect(l.window).toEqual({ left: 0, top: 0, width: 1920, height: 1080 });
    expect(l.media.width).toBeCloseTo(1920);
    expect(l.media.height).toBeCloseTo(1080 * (1080 / 810), 1);
  });

  it("survives a box or source with no size", () => {
    const l = cropLayout(undefined, { width: 0, height: 0 }, 100, 50, "fit");
    expect(l.window).toEqual({ left: 0, top: 0, width: 100, height: 50 });
    expect(l.media).toEqual({ left: 0, top: 0, width: 100, height: 50 });
  });
});

describe("cropToAspect", () => {
  it("takes the sides off something too wide", () => {
    const c = cropToAspect({ width: 1920, height: 1080 }, 1); // to square
    expect(c.left).toBeCloseTo(c.right!);
    expect(c.top ?? 0).toBe(0);
    // 1080 of 1920 kept, so 420px a side.
    expect(c.left! * 1920).toBeCloseTo(420, 0);
  });

  it("takes the top and bottom off something too tall", () => {
    const c = cropToAspect({ width: 1080, height: 1920 }, 16 / 9);
    expect(c.top).toBeCloseTo(c.bottom!);
    expect(c.left ?? 0).toBe(0);
  });

  it("does nothing when the shapes already agree", () => {
    expect(isEmptyCrop(cropToAspect({ width: 1920, height: 1080 }, 16 / 9))).toBe(true);
    // And a rounding-error mismatch is not worth cropping for.
    expect(isEmptyCrop(cropToAspect({ width: 1918, height: 1080 }, 16 / 9))).toBe(true);
  });
});

/*
The fill focus, which decides what a filled clip throws away.

The numbers here are the exporter's: fitted by height, a 900-wide source stays
900 in a 640 frame, so 260px overflow. At focus 0 the window starts at the left
edge, at 1 it ends at the right, at 0.5 it is centred — exactly ffmpeg's
crop=(iw-W)*f, which is what keeps the preview showing the part the export keeps.
*/
describe("cropLayout fill focus", () => {
  const src = { width: 900, height: 360 };

  const leftOf = (focus: [number, number]) =>
    cropLayout(undefined, src, 640, 360, "fill", focus).window.left;

  it("hangs the overflow where the focus says", () => {
    expect(leftOf([0, 0.5])).toBe(0); // keep the left edge
    expect(leftOf([1, 0.5])).toBe(-260); // keep the right edge
    expect(leftOf([0.5, 0.5])).toBe(-130); // centred
  });

  it("defaults to centred, which is what it always did", () => {
    expect(cropLayout(undefined, src, 640, 360, "fill").window.left).toBe(-130);
  });

  it("clamps a focus that has run off the end", () => {
    expect(leftOf([-3, 0.5])).toBe(0);
    expect(leftOf([9, 0.5])).toBe(-260);
  });

  // A letterboxed picture has no overflow to choose between, and offsetting it
  // would slide the picture around inside its own bars — a different feature,
  // and not one anyone asked for.
  it("ignores the focus when the clip letterboxes", () => {
    const tall = { width: 360, height: 900 };
    const centred = cropLayout(undefined, tall, 640, 360, "fit").window.left;
    expect(cropLayout(undefined, tall, 640, 360, "fit", [0, 0]).window.left).toBe(centred);
  });
});

File: backend/internal/render/crop_test.go
package render

import (
	"strings"
	"testing"

	"studio/internal/cursor"
	"studio/internal/schema"
)

func TestCropPixelsAreAlwaysEven(t *testing.T) {
	// H.264 stores chroma at half resolution: an odd width or offset has no
	// representation in 4:2:0, and ffmpeg rounds it silently rather than
	// refusing — which shows up as a preview and an export a pixel apart.
	c := &schema.Crop{Top: 0.137, Left: 0.211, Right: 0.073, Bottom: 0.019}
	x, y, w, h := c.Pixels(1913, 1077)
	for name, v := range map[string]int{"x": x, "y": y, "w": w, "h": h} {
		if v%2 != 0 {
			t.Errorf("%s = %d, want an even number", name, v)
		}
	}
	if x+w > 1913 || y+h > 1077 {
		t.Errorf("crop %d,%d %dx%d escapes a 1913x1077 frame", x, y, w, h)
	}
}

func TestCropCannotEatTheWholeFrame(t *testing.T) {
	// Two handles dragged past each other. The result must still be a picture.
	c := &schema.Crop{Top: 0.8, Bottom: 0.8, Left: 0.95, Right: 0.95}
	_, _, w, h := c.Pixels(1920, 1080)
	if w < 2 || h < 2 {
		t.Fatalf("crop collapsed to %dx%d", w, h)
	}
	if w > 1920 || h > 1080 {
		t.Fatalf("crop grew to %dx%d", w, h)
	}
}

func TestEmptyCropIsIdentity(t *testing.T) {
	for _, c := range []*schema.Crop{nil, {}, {Top: 0, Left: 0}} {
		if !c.Empty() {
			t.Errorf("%+v should be empty", c)
		}
		if cropFilter(c, 1920, 1080) != "" {
			t.Errorf("%+v produced a filter", c)
		}
		w, h := croppedDims(1920, 1080, c)
		if w != 1920 || h != 1080 {
			t.Errorf("%+v changed the dimensions to %dx%d", c, w, h)
		}
	}
}

// An asset whose dimensions were never probed must stay unknown. Turning 0x0
// into a confident 2x2 would make the prefit letterbox every such clip into a
// speck.
func TestCropLeavesUnknownDimensionsUnknown(t *testing.T) {
	w, h := croppedDims(0, 0, &schema.Crop{Top: 0.1})
	if w != 0 || h != 0 {
		t.Errorf("croppedDims(0,0) = %dx%d, want 0x0", w, h)
	}
	if cropFilter(&schema.Crop{Top: 0.1}, 0, 0) != "" {
		t.Error("cropFilter invented a crop for a source of unknown size")
	}
}

func TestCropFilterTrimsTheRequestedEdge(t *testing.T) {
	// The headline case: cut the top quarter off a screen recording.
	got := cropFilter(&schema.Crop{Top: 0.25}, 1920, 1080)
	if got != "crop=1920:810:0:270," {
		t.Errorf("cropFilter = %q", got)
	}
	w, h := croppedDims(1920, 1080, &schema.Crop{Top: 0.25})
	if w != 1920 || h != 810 {
		t.Errorf("croppedDims = %dx%d, want 1920x810", w, h)
	}
}

/*
The pointer track has to move with the picture.

A sidecar's samples are in the recorded video's own pixels. Cropping the top off
without shifting them leaves every highlight, ring and drawn cursor exactly as
far below where it belongs as the crop was deep — which reads as the cursor
effects being miscalibrated rather than as a crop that forgot something.
*/
func TestCropShiftsTheCursorTrack(t *testing.T) {
	tr := &cursor.Track{Version: 1, Clicks: true}
	tr.Video.Width, tr.Video.Height = 1920, 1080
	tr.Samples = []cursor.Sample{
		{T: 0, X: 960, Y: 540},   // centre
		{T: 10, X: 960, Y: 100},  // inside the part being cut away
		{T: 20, X: 960, Y: 1000}, // still inside after the crop
	}

	out := cropCursor(tr, &schema.Crop{Top: 0.25})
	if out.Video.Height != 810 {
		t.Fatalf("track frame = %dx%d, want 1920x810", out.Video.Width, out.Video.Height)
	}
	if len(out.Samples) != 2 {
		t.Fatalf("want the two surviving samples, got %d", len(out.Samples))
	}
	if out.Samples[0].Y != 270 {
		t.Errorf("sample not shifted: y = %d, want 270", out.Samples[0].Y)
	}
	// Clicks and timing must survive the shift, or the rings drift off the
	// presses they mark.
	if out.Samples[1].T != 20 || out.Samples[1].Y != 730 {
		t.Errorf("second sample = %+v", out.Samples[1])
	}
	// The original must not be mutated: it is read once per export and could be
	// shared between two clips of the same asset with different crops.
	if tr.Video.Height != 1080 || len(tr.Samples) != 3 {
		t.Error("cropCursor mutated the track it was given")
	}
}

func TestCropCursorIsIdentityWithoutACrop(t *testing.T) {
	tr := &cursor.Track{Version: 1}
	tr.Video.Width, tr.Video.Height = 1920, 1080
	tr.Samples = []cursor.Sample{{T: 0, X: 5, Y: 5}}
	if got := cropCursor(tr, nil); got != tr {
		t.Error("an empty crop should hand back the same track")
	}
}

func TestPrefitModes(t *testing.T) {
	// A 4:3 source in a 16:9 canvas — the shapes genuinely disagree.
	const sw, sh, w, h = 1440, 1080, 1920, 1080

	fit := prefitFilter(schema.FitContain, sw, sh, w, h)
	if !strings.Contains(fit, "decrease") || !strings.Contains(fit, "pad=") {
		t.Errorf("fit should letterbox: %q", fit)
	}
	fill := prefitFilter(schema.FitCover, sw, sh, w, h)
	if !strings.Contains(fill, "increase") || !strings.Contains(fill, "crop=1920:1080") {
		t.Errorf("fill should cover and crop: %q", fill)
	}
	if got := prefitFilter(schema.FitStretch, sw, sh, w, h); got != "" {
		t.Errorf("stretch should add nothing, leaving the plain scale: %q", got)
	}
	// Shapes that already agree need no prefit in any mode.
	for _, mode := range []string{schema.FitAuto, schema.FitContain, schema.FitCover} {
		if got := prefitFilter(mode, 1920, 1080, w, h); got != "" {
			t.Errorf("%q prefitted a matching source: %q", mode, got)
		}
	}
}

/*
The default letterboxes, whatever else is true of the clip.

It used to fill on any clip the camera was working, so that a push-in could never
reveal the bar beside the picture — and since every screen recording carries
cursor effects, that quietly cropped a quarter off any recording whose shape did
not match the canvas. The bar is kept out of frame by the pan clamp instead (see
clipBoxAt); the default's job is only to avoid throwing away picture nobody asked
to lose.
*/
func TestAutoFitLetterboxes(t *testing.T) {
	const sw, sh, w, h = 1440, 1080, 1920, 1080
	for _, name := range []string{"a plain clip", "a clip the camera works"} {
		got := prefitFilter(schema.FitAuto, sw, sh, w, h)
		if !strings.Contains(got, "decrease") {
			t.Errorf("%s should letterbox by default: %q", name, got)
		}
	}
	// An explicit choice still wins, in both directions.
	if got := prefitFilter(schema.FitCover, sw, sh, w, h); !strings.Contains(got, "increase") {
		t.Errorf("explicit fill was ignored: %q", got)
	}
	if got := prefitFilter(schema.FitContain, sw, sh, w, h); !strings.Contains(got, "decrease") {
		t.Errorf("explicit fit was ignored: %q", got)
	}
}

/*
A crop changes the shape the prefit is fitting.

This is the whole point of the feature working end to end: trim a 16:9 recording
down to a square and the clip must letterbox as a square, not go on being
treated as 16:9 and stretched. Getting this wrong is invisible in the crop
itself and shows up as the picture being subtly the wrong shape.
*/
func TestCropChangesWhatThePrefitFits(t *testing.T) {
	// 1920x1080 with half the width cropped away is 960x1080 — taller than wide.
	cw, ch := croppedDims(1920, 1080, &schema.Crop{Left: 0.25, Right: 0.25})
	if cw != 960 || ch != 1080 {
		t.Fatalf("croppedDims = %dx%d, want 960x1080", cw, ch)
	}
	seg := prefitFilter(schema.FitContain, cw, ch, 1920, 1080)
	if !strings.Contains(seg, "pad=1920:1080") {
		t.Errorf("the cropped shape was not letterboxed into the canvas: %q", seg)
	}
	// Without the crop the source already matched the canvas and needed nothing.
	if got := prefitFilter(schema.FitContain, 1920, 1080, 1920, 1080); got != "" {
		t.Errorf("uncropped source should need no prefit: %q", got)
	}
}
